cmd/syncComparator.go

```diff
@@ -54,30 +202,217 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+	lcFileName = filepath.Base(lcRelativePath)
+
+	f.sourceFolderIndex.lock.Lock()
+
+	foldermap, folderPresent := f.sourceFolderIndex.folderMap[lcFolderName]
+
+	// Do the auditing of map.
+	defer func() {
+		if folderPresent {
+			if len(f.sourceFolderIndex.folderMap[lcFolderName].indexMap) == 0 {
+				size := int64(unsafe.Sizeof(objectIndexer{}))
+				atomic.AddInt64(&f.sourceFolderIndex.totalSize, -size)
+				delete(f.sourceFolderIndex.folderMap, lcFolderName)
 			}
 		}
-	} else {
-		// purposefully ignore the error from destinationCleaner
-		// it's a tolerable error, since it just means some extra destination object might hang around a bit longer
-		_ = f.destinationCleaner(destinationObject)
+		f.sourceFolderIndex.lock.Unlock()
+	}()
+
+	// Folder Case.
+	if destinationObject.entityType == common.EEntityType.Folder() {
+		if destinationObject.isFolderEndMarker {
+			//
+			// We will see this special "end of folder" marker *after* we have seen all direct children of
+			// that directory (if any). We have the following cases:
+			// 1. destinationObject.relativePath directory does not exist in the target. In this case there won't be any children
+			//    of destinationObject.relativePath queued for us and we need to copy the directory recursively to the target.
+			// 2. destinationObject.relativePath directory exists in the target and the source directory has not "changed" since last
+			//    sync. In this case the Target traverser would have skipped enumeration of destinationObject.relativePath and hence
+			//    there won't be any children queued for us. We need to go over all its children added in sourceFolderIndex and test
+			//    them for "newness" and copy them if they are found modified after lastsync.
+			// 2. destinationObject.relativePath directory exists in the target and the source directory has "changed" since last sync.
+			//    In this case Target traverser would have enumerated the directory and queued all enumerated children for us. After all
+			//    the children it would have queued this special isFolderEndMarker object. So when we process the isFolderEndMarker object,
+			//    we would have processed all children and hence whatever is left in sourceFolderIndex for that directory are only the files
+			//    newly created in the source since last sync. We need to copy all of them to the target.
+			//
+			lcFolderName = path.Join(lcFolderName, lcFileName)
+			foldermap, folderPresent = f.sourceFolderIndex.folderMap[lcFolderName]
+			if folderPresent {
+				fmt.Printf("Finalizing directory %s (FinalizeAll=%v)\n", lcFolderName, destinationObject.isFinalizeAll)
+				f.FinalizeTargetDirectory(foldermap, destinationObject.isFinalizeAll)
+			}
+			return nil
+		} else {
+			if folderPresent {
+				sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+				// Lets check if entry in source is folder and entry on destination should be folder.
+				// Otherwise folder in source changed to file.
```

Comment:
//
                // Target also has this object and is of the same type (directory) as the source?
                // If yes, do nothing right now. When we process this directory from tqueue and finalize it,
                // that time we will sync the directory properties if needed.
                // If either target does not have object with the same name or object type is different, we
                // will need to delete it. This will be done by the f.destinationCleaner() call below.
                // The new directory will be created in FinalizeTargetDirectory() when we go over the folderIndexer
                // map entries (note that we don't remove it from the folderIndexer here).
                //